src/NServiceBus.Gateway/Gateway.cs

```diff
@@ -50,28 +51,33 @@ protected override void Setup(FeatureConfigurationContext context)
 
             ConfigureTransaction(context);
 
-            string gatewayInputAddress;
-
-            var requiredTransactionSupport = context.Settings.GetRequiredTransactionModeForReceives();
-
-            var gatewayPipeline = context.AddSatellitePipeline("Gateway", requiredTransactionSupport, PushRuntimeSettings.Default, "gateway", out gatewayInputAddress);
-
             var channelManager = CreateChannelManager(context);
 
             Func<string, IChannelSender> channelSenderFactory;
             Func<string, IChannelReceiver> channelReceiverFactory;
             RegisterChannels(context, channelManager, out channelSenderFactory, out channelReceiverFactory);
 
-            gatewayPipeline.Register(ForwardFailedGatewayMessagesToErrorQueueBehavior.StepId, b => new ForwardFailedGatewayMessagesToErrorQueueBehavior(gatewayInputAddress, b.Build<CriticalError>()), "Handles failures when sending messages through the gateway");
+            var gatewayInputAddress = context.Settings.GetTransportAddress(new LogicalAddress(context.Settings.EndpointInstanceName(), "gateway"));
 
-            var retryPolicy = context.Settings.HasSetting("Gateway.Retries.RetryPolicy") ? context.Settings.Get<Func<IncomingMessage, Exception, int ,TimeSpan>>("Gateway.Retries.RetryPolicy") : DefaultRetryPolicy.BuildWithDefaults();
+            var requiredTransactionSupport = context.Settings.GetRequiredTransactionModeForReceives();
 
-            gatewayPipeline.Register(new RetryFailedGatewayMessagesBehavior.Registration(gatewayInputAddress, retryPolicy));
-            gatewayPipeline.Register("GatewaySendProcessor", b => new GatewaySendBehavior(gatewayInputAddress, channelManager, new MessageNotifier(), b.Build<IDispatchMessages>(), context.Settings, CreateForwarder(channelSenderFactory, b.BuildAll<IDataBus>()?.FirstOrDefault()), GetConfigurationBasedSiteRouter(context)), "Processes messages to be sent to the gateway");
-            context.Pipeline.Register("RouteToGateway", b => new RouteToGatewayBehavior(gatewayInputAddress), "Reroutes gateway messages to the gateway");
+            var retryPolicy = context.Settings.Get<Func<IncomingMessage, Exception, int, TimeSpan>>("Gateway.Retries.RetryPolicy");
+
+            var sender = new GatewayMessageSender(
+                gatewayInputAddress,
+                channelManager,
+                new MessageNotifier(),
+                context.Settings.LocalAddress(),
+                GetConfigurationBasedSiteRouter(context));
+
+            context.AddSatelliteReceiver("Gateway", gatewayInputAddress, requiredTransactionSupport, PushRuntimeSettings.Default,
+                (config, errorContext) => GatewayRecoverabilityPolicy.Invoke(errorContext, retryPolicy, config),
+                (builder, messageContext) => sender.SendToDestination(messageContext, builder.Build<IDispatchMessages>(), CreateForwarder(channelSenderFactory, builder.BuildAll<IDataBus>()?.FirstOrDefault())));
```

Comment:
>The dispatcher should be stateless by design

But we need to get it from the builder so there is no other way to do it? (or did you mean the "forwarder")